### PR TITLE
Feature/day3/add code

### PR DESCRIPTION
**Pull request by jkelly282**

### CODE REVIEW

File: day3/test_day3.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import unittest
+
+import day3
+
+
+class TestDay3(unittest.TestCase):
+    def test_open_file(self):
```
Comment (GvandeSteeg) [inline day3/test_day3.py]: day3 doesn't have an `open_file` function

File: day3/test_day3.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        mylines = ["1234"]
+        self.assertNotEqual(expected, day3.parse_claims(mylines))
+
+    def test_parse_claims_type(self):
+        mylines = 43770
+        self.assertRaises(TypeError, day3.parse_claims, mylines)
+
+    def test_parse_claims_empty(self):
+        expected = [['1', '2', '3', '4']]
+        mylines = [""]
```
Comment (GvandeSteeg) [inline day3/test_day3.py]: shouldn't this throw an error? might want to update the code to test for empty input and raise an exception instead

File: day3/test_day3.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        self.assertNotEqual(expected, day3.parse_claims(mylines))
```
Comment (GvandeSteeg) [inline day3/test_day3.py]: for readability, it's clearer to use `assertEqual` if you have an `expected` output, and modify your expected output to be what you.. expect it to be

File: day3/day3.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import logging
+import re
+
+import numpy as np
+
+import day2
+
+
+def parse_claims(mylines):
+    fabric_coordinates = []
```
Comment (GvandeSteeg) [inline day3/day3.py:24]: missing docstring

File: day3/day3.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    for i in mylines:
+        # gets rid of the empty spaces
+        j = i.replace(' ', '')
+        parsed_claim = re.split('@|,|:|x', j)
+        #remove the #claim
+        fabric_coordinates.append(parsed_claim[1:5])
+    return fabric_coordinates
+
+
+def parse_fabric_list(claims):
```
Comment (GvandeSteeg) [inline day3/day3.py]: missing docstring

Comment (GvandeSteeg) [inline day3/day3.py]: missing test

Comment (GvandeSteeg) [reply day3/day3.py]: What is `claims`? Use type hints and/or docstrings to make it clearer

File: day3/day3.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        santa_fabric[i[1]:i[1] + i[3], i[0]:i[0] + i[2]] += 1
+    return np.sum(santa_fabric > 1)
+
+
+if __name__ == '__main__':
+    args = day2.parse_args()
+    try:
+        mylines = day2.open_file(args.text_strings)
+    except ValueError:
+        logging.error("File should contain lines")
```
Comment (GvandeSteeg) [inline day3/day3.py:69]: what happens if you provide an empty file? Try writing a test for that, you'd be surprised ;)

**Review by GvandeSteeg — COMMENTED** [review]
You're improving very quickly! My comments aren't codewise, but style this time. Well done!

File: day3/day3.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        j = i.replace(' ', '')
+        parsed_claim = re.split('#|@|,|:|x', j)
+        # remove the #claim
+        fabric_coordinates.append(parsed_claim[1:6])
+    return fabric_coordinates
+
+
+def parse_fabric_list(claims, size):
+    '''
+    Puts the elves claims into an array object
```
Comment (GvandeSteeg) [inline day3/day3.py]: This function seems to do more than just putting a claim into an array object, could you make it a bit more descriptive?

File: day3/day3.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    for i in mylines:
+        # gets rid of the empty spaces
+        j = i.replace(' ', '')
+        parsed_claim = re.split('#|@|,|:|x', j)
+        # remove the #claim
+        fabric_coordinates.append(parsed_claim[1:6])
+    return fabric_coordinates
+
+
+def parse_fabric_list(claims, size):
```
Comment (GvandeSteeg) [inline day3/day3.py]: Add some type hints here, along the lines of `def functionname(arg1: type1, arg2: type2) -> (returntype1, returntype2):`

File: day3/day3.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    '''
+    santa_fabric = np.zeros((size, size))
+    for i in claims:
+        # converting the list of strings to int
+        i = list(map(int, i[1:5]))
+        santa_fabric[i[1]:i[1] + i[3], i[0]:i[0] + i[2]] += 1
+    return np.sum(santa_fabric > 1), santa_fabric
+
+
+def uncontested_claim(claims, fabric_map: list) -> int:
```
Comment (GvandeSteeg) [inline day3/day3.py]: add typehint for `claim`

File: day3/day3.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    '''
+    Checks for claims in the fabric which do not overlap with other claims
+
+    :param fabric_map: an array containing the submitted elf claims for santa's fabric:
+    :return: the claim that does not conflict with another elf's claim
+    '''
+    for i in claims:
+        i = list(map(int, i))
+        a, b, c, d, e = i
```
Comment (GvandeSteeg) [inline day3/day3.py]: if you're going to unpack a list, use clearer variables than this. Either put them in the direct order with another function, and show those to `fabric_map` 
e.g. 
```python
identifier, coords = get_coords(i)
if np.all(fabric_map[coords] == 1):
    return identifier
```

File: day3/day3.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        # remove the #claim
+        fabric_coordinates.append(parsed_claim[1:6])
+    return fabric_coordinates
+
+
+def parse_fabric_list(claims, size):
+    '''
+    Puts the elves claims into an array object
+
+    :param claims: a list object made up of lists in the format [['str','str','str','str'],['str','str','str','str']]
```
Comment (GvandeSteeg) [inline day3/day3.py:39]: missing docstring for parameter `size`

File: day3/day3.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        i = list(map(int, i[1:5]))
+        santa_fabric[i[1]:i[1] + i[3], i[0]:i[0] + i[2]] += 1
+    return np.sum(santa_fabric > 1), santa_fabric
+
+
+def uncontested_claim(claims, fabric_map: list) -> int:
+    '''
+    Checks for claims in the fabric which do not overlap with other claims
+
+    :param fabric_map: an array containing the submitted elf claims for santa's fabric:
```
Comment (GvandeSteeg) [inline day3/day3.py:54]: add docstring for parameter `claim`
remember:
1. update code
2. update docstring
3. update test

File: day3/day3.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+if __name__ == '__main__':
+    args = day2.parse_args()
+    try:
+        mylines = day2.open_file(args.text_strings)
+    except ValueError:
+        logging.error("File should contain lines")
+        raise
+
+    claims = parse_claims(mylines)
+    fabric = parse_fabric_list(claims, 1000)
```
Comment (GvandeSteeg) [inline day3/day3.py]: why is 1000 hardcoded?

File: day3/test_day3.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        self.assertRaises(ValueError, day3.parse_claims, mylines)
+
+    def test_parse_fabric_list(self):
+        expected = 4
+        claims = [[1, 1, 3, 4, 4], [2, 3, 1, 4, 4]]
+        self.assertEqual(expected, day3.parse_fabric_list(claims, 8))
+
+    def test_parse_fabric_list_string(self):
+        claims = [[1, 1, 3, 4, 4], [2, 3, 1, 4, 4]]
+        self.assertRaises(TypeError, day3.parse_fabric_list, claims, "8")
```
Comment (GvandeSteeg) [inline day3/test_day3.py:35]: you're testing Python base functionality here. Instead add some information to the user why `"8"` is not acceptable. They'll get an abstract error which they might not understand

File: day3/test_day3.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+    def test_parse_fabric_list_size_dosent_matter(self):
```
Comment (GvandeSteeg) [inline day3/test_day3.py:37]: `doesnt` :P

Comment (GvandeSteeg) [reply day3/test_day3.py:37]: or `does_not`

File: day3/test_day3.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        expected = 4
+        claims = [[1, 1, 3, 4, 4], [2, 3, 1, 4, 4]]
+        self.assertEqual(expected, day3.parse_fabric_list(claims, 1000))
+
+    def test_uncontested_claims(self):
```
Comment (GvandeSteeg) [inline day3/test_day3.py:42]: I want to see at least one more test for this, where `np.all != 1`

File: day3/day3.py (hunk 9 above)
Comment (GvandeSteeg) [inline day3/day3.py]: if [line60](https://github.com/jkelly282/Advent-of-code/pull/7/files#diff-ea3be4a331efa186c31eb43abb553172R60 ) does not return a value, this function will return `None`. If that's not what you want, either have it raise an exception or return 0.
If you're okay with it returning `None`, update the typehint to `-> Optional[int]`

**Review by GvandeSteeg — CHANGES_REQUESTED** [review]
Have a bit of a close look at your docstrings, and don't get lazy with your tests ;)
apart from that, looking good

File: day3/day3.py (hunk 7 above)
Comment (jkelly282) [reply day3/day3.py]: Done !

File: day3/day3.py (hunk 13 above)
Comment (jkelly282) [reply day3/day3.py]: Fixed !

File: day3/day3.py (hunk 7 above)
Comment (GvandeSteeg) [reply day3/day3.py]: is it? the docstring hasn't changed!

File: day3/day3.py (hunk 11 above)
Comment (GvandeSteeg) [reply day3/day3.py:39]: stop marking my comments as resolved when you haven't addressed them, or at least provide arguments why you decide against it XD

**Pull request merged**